fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -491,16 +491,80 @@ struct CommitBatchContext {
 
 	std::unordered_map<uint16_t, Version> tpcvMap; // obtained from sequencer
 	std::set<uint16_t> writtenTLogs; // the set of tlog locations written to in the mutation.
-	std::set<Tag> writtenTags; // the set of tags written to in the mutation.
+	std::set<Tag> writtenTags; // final set tags written to in the batch
+	std::set<Tag> writtenTagsPreResolution; // tags written to in the batch not including any changes from the resolver.
+	bool hasMetadataMutation = false;
+	bool metadataMutationFromProxy = false;
 
 	CommitBatchContext(ProxyCommitData*, const std::vector<CommitTransactionRequest>*, const int);
 
 	void setupTraceBatch();
 
+	std::set<Tag> getWrittenTagsPreResolution();
+
 private:
 	void evaluateBatchSize();
 };
 
+std::set<Tag> CommitBatchContext::getWrittenTagsPreResolution() {
+	std::set<Tag> transactionTags;
+	std::vector<Tag> cacheVector = { cacheTag };
+	for (int transactionNum = 0; transactionNum < trs.size(); transactionNum++) {
+		int mutationNum = 0;
+		VectorRef<MutationRef>* pMutations = &trs[transactionNum].transaction.mutations;
+		for (; mutationNum < pMutations->size(); mutationNum++) {
+			auto& m = (*pMutations)[mutationNum];
+			if (m.param1.startsWith(serverTagPrefix)) {
```

Comment:
This is a metadata mutation. I wonder why this mutation is special to be included here, but others are not.

Reply:
if there is a metadata mutation we will send to all tLogs, so including the tag here won't matter. But the function should not be including it; I'll remove it.

Reply:
If there is a metadata mutation we will send to all tlogs, and that helps order the versions on all tlogs. That's fine.
But I think we still need to extract the tags affected by the system/metadata mutations (received prior to the resolution step). Without doing that, I think the version vector on the sequencer won't be updated properly and that could cause reads to return wrong results.
For example, suppose moveKeys() is moving a shard from storage server A to B. Suppose v_mk is the commit version of the moveKeys() operation. I think the proxy currently includes both tags A and B as being affected by that commit and the sequencer updates the versions of A and B to v_mk when moveKeys() commits. So any subsequent reads to B for that shard will get a commit version >= v_mk from the version vector. Since we are not extracting the tags affected by the (already received) system/metadata mutations above I think we won't update the version vector correctly.

Reply:
The final set of tags used to update the version vector is not created here. The final set (`writtenTags`) is built after we call `applyMetadataMutations`, later on. In that function, we will add in + serialize all metadata tags.

Reply:
That's correct, we will have the correct TPCV. But I think my question is how (and when) do we extract the affected tags and how (and when) do we update the version vector if there are metadata mutations in the current transaction batch (that we are trying to commit)? Pre-resolution or post-resolution?

Reply:
And, to be more precise, the proxy sends both the tlog set (to order versions on tlogs) and the tag set (to update the version vector) to the sequencer - it appears to me the correct tag set is obtained only post-resolution if there are metadata mutations but the call to the sequencer gets made pre-resolution and so the proxy won't be able to send the correct tag set to the sequencer in this case?

Reply:
sorry, I do not understand what you mean.

Reply:
Ok, that's what I was trying to understand. The work in this PR is to obtain TPCV only and we compute the affected tags the way we used to. Ok, I follow now.